internal/services/admin/server.go

```diff
@@ -204,6 +207,69 @@ func (a *AdminServiceImpl) TriggerWorkflow(ctx context.Context, req *contracts.T
 	}, nil
 }
 
+func (a *AdminServiceImpl) BulkTriggerWorkflow(ctx context.Context, req *contracts.BulkTriggerWorkflowRequest) (*contracts.BulkTriggerWorkflowResponse, error) {
+	tenant := ctx.Value("tenant").(*dbsqlc.Tenant)
+	tenantId := sqlchelpers.UUIDToStr(tenant.ID)
+
+	createContext, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancel()
+	var opts []*repository.CreateWorkflowRunOpts
+	var existingWorkflows []string
+
+	if len(req.Workflows) == 0 {
+		return &contracts.BulkTriggerWorkflowResponse{}, status.Error(codes.InvalidArgument, "no workflows provided")
+	}
+
+	if len(req.Workflows) > 1000 {
+		return nil, status.Error(codes.InvalidArgument, "maximum of 1000 workflows can be triggered at once")
+	}
+
+	opts, existingWorkflows, err := getOpts(ctx, req.Workflows, a)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(opts) == 0 {
+		if len(existingWorkflows) == 0 {
+			return &contracts.BulkTriggerWorkflowResponse{}, status.Error(codes.InvalidArgument, "no suitable new workflows provided")
+		}
+		return &contracts.BulkTriggerWorkflowResponse{WorkflowRunIds: existingWorkflows}, nil
+
+	}
+
+	workflowRunIds, err := a.repo.WorkflowRun().CreateNewWorkflowRuns(createContext, tenantId, opts)
+
+	if err == metered.ErrResourceExhausted {
+		return nil, status.Errorf(codes.ResourceExhausted, "resource exhausted: workflow run limit exceeded for tenant")
+	}
+	if err != nil {
+		return nil, fmt.Errorf("could not create workflow runs: %w", err)
+	}
+
+	for _, workflowRunId := range workflowRunIds {
+		err = a.mq.AddMessage(
+			context.Background(),
+			msgqueue.WORKFLOW_PROCESSING_QUEUE,
+			tasktypes.WorkflowRunQueuedToTask(tenantId, workflowRunId),
+		)
+
+		if err != nil {
+			return nil, fmt.Errorf("could not queue workflow run: %w", err)
```

Comment:
I feel it is probably the cleanest failure state, otherwise we need to support partial errors and a partially failed transaction, which is going to increase the complexity a lot.